crates/voicevox_core/src/asyncs.rs

```diff
@@ -0,0 +1,54 @@
+use std::{
+    io::{self, Read as _, Seek as _, SeekFrom},
+    path::Path,
+    pin::Pin,
+    task::{self, Poll},
+};
+
+use futures_io::{AsyncRead, AsyncSeek};
+
+pub(crate) trait Async: 'static {
+    async fn open_file(path: impl AsRef<Path>) -> io::Result<impl AsyncRead + AsyncSeek + Unpin>;
+}
+
+/// "async"としての責務を放棄し、すべてをブロックする。
+pub(crate) enum Unstoppable {}
+
+impl Async for Unstoppable {
```

Comment:
これがblocking用だというのが分かりづらいかも･･･？
（ @sevenc-nanashi さん的に別に分かりにくくなければ良いかも）

Rustの内部実装を考えると「即座にResultを返す」とか、「Pollingを返さない」とか？
あるいは`ForBlocking`とかでも分かりやすかったり･･･？

Reply:
PseudoAsyncとか...？

Reply:
よく考えたらスレッドプール利用も"unstoppable"と言えるし、そもそもあまり適切ではなかったですね。

私からの提案としては、`SingleTask`はどうでしょうか？

```rs
/// 完全にシングルタスクで動くことを仮定する。
///
/// [ブロッキング版API]用。
///
/// `async`の中でブロッキング操作を直接行う。
///
/// [ブロッキング版API]: crate::blocking
pub(crate) enum SingleTask {}
```

```rs
/// [blocking]クレートで駆動する。
///
/// [非同期版API]用。
///
/// [blocking]: https://docs.rs/crate/blocking
/// [非同期版API]: crate::tokio
pub(crate) enum BlockingThreadPool {}
```

Reply:
難しいところですが、何をしてなぜ存在してどう実現するのか想像できないのが課題なわけで、どう頑張っても名前だけで表現するのは無理だから、ドキュメントでカバーするのが良さそう！
なので、名前は頑張りすぎなくてもよく、代わりにちょっとドキュメント書く感じですかねぇ。

SingleTaskや「完全にシングルタスクで動く」はどういうことかわからないかも。（asyncは最初からシングルタスクなはず…？）

擬似Asyncは可能ならもう少し意味を込められるとかっこよさそう。

個人的には制御を返さないことが主点だと思うので、BlockingAsyncだとかBlockingだとかもありに感じます。

ま、とはいえコメントドキュメントでわかれば良いので、どれでも良さそう！
特に決まらなければSingleTaskで良さそう！

Reply:
> SingleTaskや「完全にシングルタスクで動く」はどういうことかわからないかも。（asyncは最初からシングルタスクなはず…？）

async文脈での"task"という意味で通らないかなーと思ってます。「シングルタスク」という日本語表現は避けた方がいい感じでしょうかね？ちなみに"task"の定義は多分こんな感じです:

### Asynchronous Programming in Rust (async-book)での定義

> Each time a future is polled, it is polled as part of a "task". Tasks are the top-level futures that have been submitted to an executor.

多分これが公式の原義。

### Tokioでの定義

> ## What are Tasks?
>
> A _task_ is a light weight, non-blocking unit of execution. A task is similar to an OS thread, but rather than being managed by the OS scheduler, they are managed by the [Tokio runtime][rt]. Another name for this general pattern is [green threads]. If you are familiar with [Go's goroutines], [Kotlin's coroutines], or [Erlang's processes], you can think of Tokio's tasks as something similar.

[rt]: https://docs.rs/tokio/1/tokio/runtime/index.html
[green threads]: https://en.wikipedia.org/wiki/Green_threads
[Go's goroutines]: https://tour.golang.org/concurrency/1
[Kotlin's coroutines]: https://kotlinlang.org/docs/reference/coroutines-overview.html
[Erlang's processes]: http://erlang.org/doc/getting_started/conc_prog.html#processes

### async-stdでの定義

> ## The task model
>
> An executing asynchronous Rust program consists of a collection of native OS threads, on top of which multiple stackless coroutines are multiplexed. We refer to these as "tasks".  Tasks can be named, and provide some built-in support for synchronization.

### async-taskでの定義（= smolとblockingでの定義）

> Task abstraction for building executors.
>
> To spawn a future onto an executor, we first need to allocate it on the heap and keep some state attached to it. The state indicates whether the future is ready for polling, waiting to be woken up, or completed. Such a stateful future is called a *task*.

Reply:
```diff
-/// 完全にシングルタスクで動くことを仮定する。
+/// エグゼキュータが非同期処理の並行実行をしないことを仮定する。
 ///
 /// [ブロッキング版API]用。
 ///
 /// `async`の中でブロッキング操作を直接行う。
 ///
 /// [ブロッキング版API]: crate::blocking
 pub(crate) enum SingleTask {}
```

（rust-lang-jaが[JISにならって長音記号を省く派だった](https://rust-lang-jp.zulipchat.com/#narrow/stream/333000-translation/topic/.E5.AF.BE.E8.A8.B3.E8.A1.A8.E3.81.AB.E3.81.A4.E3.81.84.E3.81.A6/near/384884789)ため「エグゼキュータ」にしてみました）

Reply:
なるほどです！！

`SingleTask`は一番いい答えではないかもですが、ドキュメントがしっかりしていればありだと思ってます。
（`Single`が特に何も意味してない。かといって別の意味と衝突するわけでもないので、あり。）
あとは即時実行`ImmediateExecution`(Task)とか？

とりあえずドキュメントの内容は伝わるので良さそう！
一番最初に注意事項が書いてあって一般的ではないので、注意事項だとわかるように`WARNING: エグゼキュータが非同期タスクの並行実行をしないことを仮定する。`とかにすると良いかなと！

Reply:
[`18991c1` (#830)](https://github.com/VOICEVOX/voicevox_core/pull/830/commits/18991c1ed5d8e96b37c38b969d3315b012de322f)
<code>SingleTask**ed**</code>にしてみました。"single-threaded"ほど一般的ではありませんが、意味が通じるのはこっちの方かなと。あとdocstringもちょっと変えました。

ChatGPT 4oに聞くと、まあこんな感じ…
<https://chatgpt.com/share/cdae540e-5751-43a5-a1fb-ac1f17d6a1b8>

Reply:
名前決まった＆ドキュメントで案内できてるので、良さそう！resolveで･･･！